// appium-ios-simulator/build/lib/extensions/applications.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _lodash = _interopRequireDefault(require("lodash"));
var _path = _interopRequireDefault(require("path"));
var _support = require("@appium/support");
var _logger = _interopRequireDefault(require("../logger"));
var _bluebird = _interopRequireDefault(require("bluebird"));
var _asyncbox = require("asyncbox");
const extensions = {};
extensions.installApp = async function installApp(app) {
  return await this.simctl.installApp(app);
};
extensions.getUserInstalledBundleIdsByBundleName = async function getUserInstalledBundleIdsByBundleName(bundleName) {
  const appsRoot = _path.default.resolve(this.getDir(), 'Containers', 'Bundle', 'Application');
  const infoPlists = await _support.fs.glob('*/*.app/Info.plist', {
    cwd: appsRoot,
    nosort: true,
    strict: false,
    absolute: true
  });
  if (_lodash.default.isEmpty(infoPlists)) {
    return [];
  }
  const bundleInfoPromises = [];
  for (const infoPlist of infoPlists) {
    bundleInfoPromises.push((async () => {
      try {
        return await _support.plist.parsePlistFile(infoPlist);
      } catch (ign) {}
    })());
  }
  const bundleInfos = (await _bluebird.default.all(bundleInfoPromises)).filter(_lodash.default.isPlainObject);
  const bundleIds = bundleInfos.filter(({
    CFBundleName
  }) => CFBundleName === bundleName).map(({
    CFBundleIdentifier
  }) => CFBundleIdentifier);
  if (_lodash.default.isEmpty(bundleIds)) {
    return [];
  }
  _logger.default.debug(`The simulator has ${_support.util.pluralize('bundle', bundleIds.length, true)} which ` + `have '${bundleName}' as their 'CFBundleName': ${JSON.stringify(bundleIds)}`);
  return bundleIds;
};
extensions.isAppInstalled = async function isAppInstalled(bundleId) {
  try {
    const appContainer = await this.simctl.getAppContainer(bundleId);
    return appContainer.endsWith('.app');
  } catch (err) {
    try {
      const info = await this.simctl.appInfo(bundleId);
      return info.includes('ApplicationType');
    } catch (e) {
      return false;
    }
  }
};
extensions.removeApp = async function removeApp(bundleId) {
  await this.simctl.removeApp(bundleId);
};
extensions.launchApp = async function launchApp(bundleId, opts = {}) {
  await this.simctl.launchApp(bundleId);
  const {
    wait = false,
    timeoutMs = 10000
  } = opts;
  if (!wait) {
    return;
  }
  try {
    await (0, _asyncbox.waitForCondition)(async () => await this.isAppRunning(bundleId), {
      waitMs: timeoutMs,
      intervalMs: 300
    });
  } catch (e) {
    throw new Error(`App '${bundleId}' is not runnning after ${timeoutMs}ms timeout.`);
  }
};
extensions.terminateApp = async function terminateApp(bundleId) {
  await this.simctl.terminateApp(bundleId);
};
extensions.isAppRunning = async function isAppRunning(bundleId) {
  return (await this.ps()).some(({
    name
  }) => name === bundleId);
};
extensions.scrubApp = async function scrubApp(bundleId) {
  const appDataRoot = await this.simctl.getAppContainer(bundleId, 'data');
  const appFiles = await _support.fs.glob('**/*', {
    cwd: appDataRoot,
    nosort: true,
    strict: false,
    nodir: true,
    absolute: true
  });
  _logger.default.info(`Found ${appFiles.length} ${bundleId} app ${_support.util.pluralize('file', appFiles.length, false)} to scrub`);
  if (_lodash.default.isEmpty(appFiles)) {
    return;
  }
  try {
    await this.terminateApp(bundleId);
  } catch (ign) {}
  await _bluebird.default.all(appFiles.map(p => _support.fs.rimraf(p)));
};
var _default = extensions;
exports.default = _default;
