// appium-ios-simulator/build/lib/extensions/settings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("source-map-support/register");
var _lodash = _interopRequireDefault(require("lodash"));
var _defaultsUtils = require("../defaults-utils");
var _bluebird = _interopRequireDefault(require("bluebird"));
const extensions = {};
extensions.setReduceMotion = async function setReduceMotion(reduceMotion) {
  return await this.updateSettings('com.apple.Accessibility', {
    ReduceMotionEnabled: Number(reduceMotion)
  });
};
extensions.setReduceTransparency = async function setReduceTransparency(reduceTransparency) {
  return await this.updateSettings('com.apple.Accessibility', {
    EnhancedBackgroundContrastEnabled: Number(reduceTransparency)
  });
};
extensions.updateSettings = async function updateSettings(domain, updates) {
  if (_lodash.default.isEmpty(updates)) {
    return false;
  }
  const argChunks = (0, _defaultsUtils.generateDefaultsCommandArgs)(updates);
  await _bluebird.default.all(argChunks.map(args => this.simctl.spawnProcess(['defaults', 'write', domain, ...args])));
  return true;
};
extensions.setAppearance = async function setAppearance() {
  throw new Error(`Xcode SDK '${this.xcodeVersion}' is too old to set UI appearance`);
};
extensions.getAppearance = async function getAppearance() {
  throw new Error(`Xcode SDK '${this.xcodeVersion}' is too old to get UI appearance`);
};
extensions.configureLocalization = async function configureLocalization() {
  throw new Error(`Xcode SDK '${this.xcodeVersion}' is too old to configure the Simulator locale`);
};
var _default = extensions;
exports.default = _default;
